### DISCUSSION TITLE
Creating Custom Renderers

### DISCUSSION THREAD

**Discussion by S-Varunn** [jfree/jfreechart#360] [Q&A]
I have a requirement where I need to give the bars in a bar chart rounded corners. How do I do this?

**Comment by trashgod** [comment]
You'll likely need a custom [`BarPainter`](https://www.jfree.org/jfreechart/javadoc/org/jfree/chart/renderer/category/BarPainter.html). You might try extending one such as `StandardBarPainter `, overriding `paintBar()` and substituting a `RoundRectangle2D` in the call to `fill`.

**Reply by S-Varunn** [reply]
Yes I was able to achieve this by the method you have specified. 
After doing that I realised that both parts of the rectangle the top and bottom gets a rounded corner. I only need the top part of the bar to have rounded corner. 
Is it possible to achieve this using Path2D? 
```
public class TopRoundedRectangle2D extends Path2D.Float {
    public TopRoundedRectangle2D(double x,double y,double width, double height, double radius) {
        moveTo(x, height);
        lineTo(x, radius);
        curveTo(x, y, x, y, radius, y);
        lineTo(height-radius, y);
        moveTo(height-radius,y);
        curveTo(height, y, height, y,width, radius);
        lineTo(height,width);
        lineTo(x,height);
        closePath();
    }
}
```
This draws the rectangle as per my requirement. However the BarPainter accepts only RectangleShape. It is possible to achieve this without changing the source code? I need another method with Path2D signature.

**Comment by trashgod** [comment]
You could try adjusting the graphics context's `Clip`, but a custom `RectangularShape` may be required. Also look at `Chart2D`, seen [here](https://stackoverflow.com/a/28519474/230513).

**Reply by trashgod** [reply]
Also consider adding an `Ellipse2D` having  a diameter matching the bar's width. Render it at the top of the bar. Shorten the bar by the length of the ellipse's semi-minor axis.
